crates/walrus-service/src/node/checkpoint.rs

```diff
@@ -0,0 +1,820 @@
+                        );
+                    }
+                };
+            }
+
+            _ = tokio::time::sleep(timeout_duration) => {
+                let mut status_guard = status.lock().expect("Failed to lock status");
+                *status_guard = TaskStatus::Timeout;
+            }
+        }
+    }
+
+    /// Cancel the task.
+    pub async fn cancel(&self) {
+        if let Ok(mut status_guard) = self.status.lock() {
+            *status_guard = TaskStatus::Cancelled;
+            self.handle.abort();
+        }
+    }
+
+    /// Get the status of the task.
+    pub fn get_status(&self) -> TaskStatus {
+        let state = self.status.lock().unwrap();
+        state.clone()
+    }
+}
+
+/// This enum defines the requests that can be sent to the checkpoint manager.
+#[derive(Debug)]
+pub enum CheckpointRequest {
+    /// Create a checkpoint.
+    CreateCheckpoint {
+        /// The response channel.
+        response: tokio::sync::oneshot::Sender<TaskResult<CheckpointError>>,
+        /// The directory to create the checkpoint in.
+        checkpoint_dir: PathBuf,
+        /// Delay before creating the checkpoint.
+        delay: Option<time::Duration>,
+    },
+    /// Get the status of the current task.
+    GetStatus {
+        /// The response channel.
+        response: tokio::sync::oneshot::Sender<TaskStatus>,
+    },
+    /// Cancel the current task.
+    CancelBackup {
+        /// The response channel.
+        /// returns true if the backup was canceled.
+        response: tokio::sync::oneshot::Sender<bool>,
+    },
+}
+
+/// Manages the creation/cleanup of checkpoints.
+#[derive(Debug)]
+pub struct CheckpointManager {
+    /// Driver for handling checkpoint requests.
+    execution_loop: JoinHandle<Result<(), CheckpointError>>,
+    /// A simple loop that schedules checkpoint creation at a fixed interval.
+    schedule_loop_handle: Option<JoinHandle<Result<(), CheckpointError>>>,
+    /// Cancellation token.
+    cancel_token: CancellationToken,
+    /// Channel to send commands to the checkpoint manager.
+    command_tx: tokio::sync::mpsc::Sender<CheckpointRequest>,
+    /// The configuration.
+    config: CheckpointConfig,
+}
+
+impl CheckpointManager {
+    /// Initial delay before first checkpoint creation, to avoid resource contention.
+    const CHECKPOINT_CREATION_INITIAL_DELAY: time::Duration = time::Duration::from_secs(15 * 60);
+    /// Delay between checkpoint creation retries.
+    const CHECKPOINT_CREATION_RETRY_DELAY: time::Duration = time::Duration::from_secs(300);
+    /// Default timeout for tasks.
+    const DEFAULT_TASK_TIMEOUT: StdDuration = time::Duration::from_secs(60 * 60);
+
+    /// Create a new checkpoint manager for RocksDB.
+    pub async fn new(db: Arc<RocksDB>, config: CheckpointConfig) -> Result<Self, CheckpointError> {
+        if let Some(checkpoint_dir) = config.checkpoint_dir.as_ref() {
+            Self::create_checkpoint_dir_if_not_exists(checkpoint_dir)?;
```

Comment:
What happens if `config.checkpoint_dir` is None? Will it fail later due to the absence of a checkpoint dir?